numpy/core/src/multiarray/item_selection.c

```diff
@@ -787,255 +787,287 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
             goto fail;
         }
+    }
 
-        while (size--) {
-            _unaligned_strided_byte_copy(buffer, (npy_intp) elsize, it->dataptr,
-                                         astride, N, elsize);
-            if (swap) {
-                _strided_byte_swap(buffer, (npy_intp) elsize, N, elsize);
-            }
-            if (part == NULL) {
-                if (sort(buffer, N, op) < 0) {
-                    PyDataMem_FREE(buffer);
-                    goto fail;
+    while (size--) {
+        char *bufptr = it->dataptr;
+
+        if (needcopy) {
+            if (hasrefs) {
+                /*
+                 * For dtype's with objects, copyswapn Py_XINCREF's src
+                 * and Py_XDECREF's dst. This would crash if called on
+                 * an unitialized buffer, or leak a reference to each
+                 * object if initialized.
+                 *
+                 * So, first do the copy with no refcounting...
+                 */
+                _unaligned_strided_byte_copy(buffer, elsize,
+                                             it->dataptr, astride, N, elsize);
+                /* ...then swap in-place if needed */
+                if (swap) {
+                    copyswapn(buffer, elsize, NULL, 0, N, swap, op);
                 }
             }
             else {
-                npy_intp pivots[NPY_MAX_PIVOT_STACK];
-                npy_intp npiv = 0;
-                npy_intp i;
-                for (i = 0; i < nkth; i++) {
-                    if (part(buffer, N, kth[i], pivots, &npiv, op) < 0) {
-                        PyDataMem_FREE(buffer);
-                        goto fail;
-                    }
-                }
+                copyswapn(buffer, elsize, it->dataptr, astride, N, swap, op);
             }
-            if (swap) {
-                _strided_byte_swap(buffer, (npy_intp) elsize, N, elsize);
+            bufptr = buffer;
+        }
+        else if (swap) {
+            /* If the array is swapped but contiguous, unswap in-place */
+            copyswapn(it->dataptr, astride, NULL, 0, N, swap, op);
+        }
+
+        if (part == NULL) {
+            ret = sort(bufptr, N, op);
+            if (ret < 0 || (hasrefs && PyErr_Occurred())) {
+                ret = -1;
+                goto fail;
             }
-            _unaligned_strided_byte_copy(it->dataptr, astride, buffer,
-                                         (npy_intp) elsize, N, elsize);
-            PyArray_ITER_NEXT(it);
         }
-        PyDataMem_FREE(buffer);
-    }
-    else {
-        while (size--) {
-            if (part == NULL) {
-                if (sort(it->dataptr, N, op) < 0) {
+        else {
+            npy_intp pivots[NPY_MAX_PIVOT_STACK];
+            npy_intp npiv = 0;
+            npy_intp i;
+            for (i = 0; i < nkth; ++i) {
+                ret = part(bufptr, N, kth[i], pivots, &npiv, op);
+                if (ret < 0 || (hasrefs && PyErr_Occurred())) {
```

Comment:
This line probably deserves some explaining... The way `dtype->f->compare` is set up for object dtype's, under Python3, when comparing two uncomparable objects it returns a 0, i.e. considers them to be equal, so in principle the sorting continues all the way, although the result will be crap. So after sorting each 1D array, if our type has refs and hence the GIL has not been released, we check whether an error was raised, and return early if it is the case.